persist: correctly handle new encoding versions

ruchirK opened the pull request:
### Motivation

This PR teaches the persistence codebase to clear out all previously persisted data if the previously persisted data was persisted using a different version of the binary encoding than we can currently decode. We are able to use such a primitive strategy (no forwards or backwards compat) because currently we are only persisting data that we can delete at any point in time.

### Tips for reviewer

This change is broken out into three logical commits that teach the `list_keys` method to only emit keys populated by users (and not any internal keys populated by the Blob implementor), add structured errors for `Codec`, and then a third commit that adds the "delete on version-mismatch" logic + tests.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR adds a release note for any
  [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

ruchirK commented on src/persist-types/src/error.rs:
```diff
@@ -0,0 +1,59 @@
+
+/// A Codec related error.
+#[derive(Debug, Clone)]
+pub enum CodecError {
```
I went back and forth on whether this should be called `Error` and imported in as `CodecError` where needed

petrosagg commented on src/persist/src/file.rs:
```diff
@@ -419,6 +419,10 @@ impl Blob for FileBlob {
             if let Some(name) = file_name {
                 let name = name.to_str();
                 if let Some(name) = name {
+                    // Exclude the lockfile from the user's list of keys.
```
This could be `if name == Some(Self::LOCKFILE_PATH)`

ruchirK replied:
I might be missing something, but I think this needs to unwrap the option separate from checking whether it matches the lockfile because we use the `name` bound in `if let Some(name) = name {` on line 426.

More concretely, to do it the way you suggest I think I have to do something like this, or something more verbose?

```
diff --git a/src/persist/src/file.rs b/src/persist/src/file.rs
index 70d980455..303aef50b 100644
--- a/src/persist/src/file.rs
+++ b/src/persist/src/file.rs
@@ -418,13 +418,12 @@ impl Blob for FileBlob {
             let file_name = path.file_name();
             if let Some(name) = file_name {
                 let name = name.to_str();
-                if let Some(name) = name {
+                if let Some(Self::LOCKFILE_PATH) = name {
                     // Exclude the lockfile from the user's list of keys.
-                    if name == Self::LOCKFILE_PATH {
-                        continue;
-                    }
-                    ret.push(name.to_owned());
+                    continue;
                 }
+
+                ret.push(name.expect("known to be valid utf-8").to_string());
             }
         }
```

Of course, let me know if I'm being silly and missing something

petrosagg replied:
no, you're right. sorry for the silly suggestion :sweat_smile:

petrosagg commented on src/persist/src/s3.rs:
```diff
@@ -315,6 +315,11 @@ impl Blob for S3Blob {
                         if let Some(key) = key.strip_prefix(&prefix) {
+                            // Exclude the system generated lock key from the user's
+                            // list of keys.
+                            if key == Self::LOCKFILE_KEY {
```
Same, could be `if key.strip_prefix(&prefix) == Some(Self::LOCKFILE_KEY)`

petrosagg commented on src/dataflow-types/src/errors.rs:
```diff
@@ -40,8 +41,9 @@ impl Codec for DecodeError {
-        let decoded = serde_json::from_slice(buf).map_err(|e| format!("Decoding error: {}", e))?;
+    fn decode<'a>(buf: &'a [u8]) -> Result<Self, CodecError> {
+        let decoded = serde_json::from_slice(buf)
+            .map_err(|e| CodecError::from(format!("Decoding error: {}", e)))?;
```
If you provide an `impl From<serde_json::Error> for CodecError` implementation you'll be able to just do `serde_json::from_slice(buf)?;` here. This is because the `?` operator automatically runs `err.into()` on the returned errors so that you can easily use structured errors like you're doing here.

danhhz replied:
I think this pattern is nice and we use it at times in persist, but it shouldn't be applied blindly. In this case, making the change added a dep on serde_json to the persist_types crate, which is bad dep hygene (the impl just converts it into a string; it'd be a different story if CodecError kept the structured serde_json::Error and the dep was avoidable).

The other instances of this review comment are less objectionable since the errors in question in those are stdlib ones.

petrosagg commented on src/dataflow/src/source/mod.rs:
```diff
@@ -405,20 +406,28 @@ impl Codec for SourceTimestamp {
-                let pid =
-                    i32::from_le_bytes(<[u8; 4]>::try_from(slice).map_err(|err| err.to_string())?);
+                let pid = i32::from_le_bytes(
+                    <[u8; 4]>::try_from(slice).map_err(|err| CodecError::from(err.to_string()))?,
```
Same here for avoiding `map_err` with a `impl From<std::array::TryFromSliceError> for CodecError` implementation

petrosagg commented on src/dataflow/src/source/mod.rs:
```diff
@@ -405,20 +406,28 @@ impl Codec for SourceTimestamp {
         let slice = &buf[offset_start..(offset_start + 8)];
-        let offset = i64::from_le_bytes(<[u8; 8]>::try_from(slice).map_err(|err| err.to_string())?);
+        let offset = i64::from_le_bytes(
+            <[u8; 8]>::try_from(slice).map_err(|err| CodecError::from(err.to_string()))?,
```
Same here for avoiding `map_err` with a `impl From<std::array::TryFromSliceError> for CodecError` implementation

petrosagg commented on src/dataflow/src/source/mod.rs:
```diff
@@ -438,9 +447,9 @@ impl Codec for AssignedTimestamp {
+    fn decode<'a>(buf: &'a [u8]) -> Result<Self, CodecError> {
         Ok(AssignedTimestamp(u64::from_le_bytes(
-            <[u8; 8]>::try_from(buf).map_err(|err| err.to_string())?,
+            <[u8; 8]>::try_from(buf).map_err(|err| CodecError::from(err.to_string()))?,
```
Same here for avoiding `map_err` with a `impl From<std::array::TryFromSliceError> for CodecError` implementation

petrosagg commented on src/persist-types/src/codec_impls.rs:
```diff
@@ -45,8 +49,8 @@ impl Codec for String {
-    fn decode<'a>(buf: &'a [u8]) -> Result<Self, String> {
-        String::from_utf8(buf.to_owned()).map_err(|err| err.to_string())
+    fn decode<'a>(buf: &'a [u8]) -> Result<Self, CodecError> {
+        String::from_utf8(buf.to_owned()).map_err(|err| CodecError::from(err.to_string()))
```
Same here for avoiding `map_err` with a `impl From< std::str::Utf8Error> for CodecError` implementation

petrosagg commented on src/persist-types/src/error.rs:
```diff
@@ -0,0 +1,59 @@
+#[derive(Debug, Clone)]
+pub enum CodecError {
+    /// An unstructured codec related error.
+    String(String),
```
Should this variant be called `Unstructured` like `CoordError::Unstructured`?

ruchirK replied:
I'm following the convention used in `persist::error::Error` which calls the unstructured variant `String`. No huge preference one way or the other

petrosagg commented on src/persist-types/src/error.rs:
```diff
@@ -0,0 +1,59 @@
+}
+
+// Hack so we can debug_assert_eq against Result<(), CodecError>.
+impl PartialEq for CodecError {
```
maybe I'm missing something obvious, by can't this be auto-derived?

ruchirK replied:
nope, that was an oversight on my part

petrosagg commented on src/persist/src/operators/source.rs:
```diff
@@ -114,8 +114,8 @@ where
-                                    let k = K::decode(&k);
-                                    let v = V::decode(&v);
+                                    let k = K::decode(&k).map_err(|e| e.to_string());
+                                    let v = V::decode(&v).map_err(|e| e.to_string());
```
our `ore` create provides a `Result::map_err_to_string()` utility function that you can use here

petrosagg reviewed (commented):
I left a bunch of comments on implementing `From<Foo> for CodecError` as suggestions but feel free to just add a TODO comment if you're rather not spend time on this in this PR.

I'm not hitting approve cause I'm not super confident on how persist works, I'll leave this to someone from the persist team

ruchirK commented:
@petrosagg thanks for the super quick and thorough review!

danhhz commented on src/persist/src/indexed/cache.rs:
```diff
@@ -293,14 +294,30 @@ impl<B: Blob> BlobCache<B> {
+
+        let meta = match BlobMeta::decode(&bytes) {
+            Ok(meta) => meta,
+            Err(CodecError::InvalidEncodingVersion(_)) => {
```
This isn't a great spot for this check. I think the contract is that it should happen at most once at startup (and probably only in the production startup codepath). In theory, this should only get hit once as written (b/c it'll get written back out with the current version), but for something as scary as deleting all a user's persisted data, it's not as dead simple to reason about as I'd like.

Concretely, I think it should live in `coord::persistcfg::PersistConfig::Init`

ruchirK replied:
Agree that this is not a good spot for the check and relies way too much on the incidental fact that this function currently only gets called during indexed initialization. That was an oversight on my part.

Let me noodle on where to put this check -- I sort of think maybe `runtime::start` is a good place for it (I feel weird putting it totally outside of the indexed / runtime code because that feels a little abstraction breaking-y)

danhhz replied:
On the contrary, I picked `coord::persistcfg::PersistConfig::Init` precisely because this hack itself is abstraction breaking (and temporary). IMO src/persist isn't the right place for it

danhhz commented on src/persist/src/indexed/mod.rs:
```diff
@@ -1533,4 +1533,40 @@ mod tests {
     }
+
+    #[test]
+    fn clear_indexed_on_invalid_encoding() -> Result<(), Error> {
```
There's three relevant cases for us to test
- the version in storage is less than current meta: we should _delete_ the data and then start
- the version in storage matches the current meta: we should _keep_ the data and start
- the version in storage is greater than the current meta: we should _keep_ the data and _not start mz_. (consider the case where we've finished productionizing things. we definitely don't want starting an old version of mz to delete production data)

if I was thinking ahead, I'd have started at version 1 so you could write this test, but I didn't. you'll need to bump to version 1 as part of this PR so you can test the less case

ruchirK replied:
So, I didn't do a separate check for if the version is > current because, while I totally agree this is something we'll want when we expose persistent user tables or persistent kafka sources as beta it's less clear to me that we need it right now. The only data currently being persisted is the metrics stuff + experimental user table persistence, neither of which external users care about. On the other hand, users do sometimes cycle between several different versions of Materialize to figure out root cause of some issue (saw this last week with a user going between 0.9.6 - 0.9.10). I'm not sure if adding this sort of check at the moment, which could then require the user to manually clear out `mzdata/persist` is worth it? I guess another implicit assumption I have here is that we won't release anything to beta for a few weeks which is maybe not true.

I don't have a huge preference here, wdyt?

danhhz replied:
I'm sorry, I don't follow. All three cases I outlined above are relevant and necessary from the very moment we introduce this. The actual data being written by persist right now is not relevant to that.

Here's why: say this merges as-is and then goes out in v0.9.12. Then in the future someone is using v0.x.y, where we've started to promise backward compatibility/that we are a source of record. If after using mz for a while, if they accidentally start up on v0.9.12, it will delete all their data. That's totally unacceptable, even if the scenario is contrived.

> I'm not sure if adding this sort of check at the moment, which could then require the user to manually clear out mzdata/persist is worth it?

It only refuses to delete the data if you're starting up on an old version, which we explictly don't support right now anyway. https://materialize.com/docs/versions/#upgrading

ruchirK replied:
Ah yeah thats a totally fair point that I didn't consider.

ruchirK replied:
> It only refuses to delete the data if you're starting up on an old version, which we explictly don't support right now anyway.

Users end up starting on an older version when they're debugging issues to try to isolate which version led to the issue. That's the only (semi) legit time I've seen people starting up on a old version, and I think the proposed change could make some of that debugging harder, but I absolutely think its worth it given the concern raised of "what about when we say we're a system of record".

danhhz replied:
Oh to be clear, supporting version rollbacks is IMO table stakes for being a production system. I fully expect we'll have to change this policy at some point. However, given that this is the policy as it stands right now, we may as well take advantage of it

danhhz reviewed (commented):
I hesitate to even make this comment since it's a such a big change to the PR, but I'm not convinced that we should introduce the complexity of a structured error for this. An alternative that I think I like more is to introduce a method somewhere that, given a blob impl, returns the meta version and basing things on that (possibly factoring a bit of BlobMeta::decode out to call from this new method). Happy to discuss offline if you feel strongly about this approach

danhhz commented:
Sorry if it seems like I'm being more particular than usual on this PR. It's just that "automatically delete all of a user's data" is a pretty important thing to get right

ruchirK commented:
> Sorry if it seems like I'm being more particular than usual on this PR. It's just that "automatically delete all of a user's data" is a pretty important thing to get right

Not at all. Thank you for the thorough review. I put up a new pr that I think addresses the concerns highlighted here https://github.com/MaterializeInc/materialize/pull/8913 and if that seems to work better we can just close this and go with that instead?

ruchirK commented:
Closing this in favor of #8913
